statescript/executor.go

```diff
@@ -179,8 +212,95 @@ func execute(name string, timeout time.Duration) int {
 	return 0
 }
 
+// RetryLaterContext holds the information needed for retry-script storage
+type RetryLaterContext struct {
+	State          string        `json:"state"`
+	LatestExecTime time.Time     `json:"latestexectime"`
+	TotDuration    time.Duration `json:"totscriptduration"`
+}
+
+func (rtctx *RetryLaterContext) MarshalBinary() (data []byte, err error) {
+	data, err = json.Marshal(rtctx)
+	return
+}
+
+func (rtctx *RetryLaterContext) UnmarshalBinary(data []byte) error {
+
+	if err := json.Unmarshal(data, rtctx); err != nil {
+		return err
+	}
+
+	return nil
+
+}
+
+// if script has been tried previously, context data will be updated, if not
+// the newctx will simply be stored.
+func getRetryLaterContext(db *store.DBStore, scriptName string, newctx *RetryLaterContext) (*RetryLaterContext, error) {
+
+	ctx := &RetryLaterContext{}
+	if err := db.Get(scriptName, ctx); os.IsNotExist(err) {
+		err = db.Update(scriptName, newctx)
+		if err != nil {
+			return nil, err
+		}
+		return newctx, nil
+	} else if err != nil {
+		return nil, err
+	}
+
+	newctx.TotDuration += ctx.TotDuration
+
+	return newctx, nil
+
+}
+
+func (l *Launcher) handleRetryLaterError(db *store.DBStore, scriptName string, execState string, execTime time.Time) error {
+
+	elapsedTime := time.Since(execTime)
+
+	ctx, err := getRetryLaterContext(db, scriptName, &RetryLaterContext{execState, execTime, elapsedTime})
+	if err != nil {
+		return err
+	}
+
+	if err = db.Update(scriptName, ctx); err != nil {
+		return err
+	}
+
+	if ctx.TotDuration <= l.getRetryTimeout() {
+
+		log.Debugf("%s spent %s - will retry later", scriptName, ctx.TotDuration.String())
+
+		time.Sleep(l.getRetryInterval())
+
+		return ErrRetryLater
+	}
+
+	// since the statemachine will reset on error, give the script a new timeslot
+	if err = db.Remove(scriptName); err != nil {
+		log.Errorf("Failed to give the script %s a new timeslot", scriptName)
+	}
+
+	return errors.Errorf("statescript: error - retry time limit exceeded for %s", scriptName)
+
+}
+
 func (l Launcher) ExecuteAll(state, action string, ignoreError bool) error {
+
+	db := store.NewDBStore("/data/mender")
+	if db == nil {
+		return errors.New("failed to open database")
+	}
+	defer db.Close()
+
+	return l.executeAll(db, state, action, ignoreError)
+}
+
+func (l Launcher) executeAll(db *store.DBStore, state, action string, ignoreError bool) error {
```

Comment:
@pasinskim I originally split this up to do make testing easier. For now I have only unit tested my methods (which is pretty unflexible, so I might remove them, and replace it with some different tests). Still the main functionality here should be pretty straight forward, now it is just baked into the main execution path, and pauses it when a script requires a retry.